### PR TITLE
Use UNIXSocket to pass file descriptor

### PR DESCRIPTION
**Pull request by xu-cheng**
This is a more standard way to pass fd in UNIX world. At the same time, it helps to remove a few hacks and simplifies the code in the sandbox.(I can remove a `fork` block because `exec` can now be replaced with `system`)

cc @jacknagel

### CODE REVIEW

**Comment by jacknagel** [comment]
Which fork do you want to remove? We can't remove the fork in the installer, as some features depend on the formula being re-loaded with the correct ARGV set up.

**Comment by jacknagel** [comment]
Nevermind, I think I misunderstood, that should still work.

**Comment by xu-cheng** [comment]
> Which fork do you want to remove? 

The fork in my current implement `sandbox`. [Here](https://github.com/Homebrew/homebrew/pull/38361/files#diff-35b8c3e4d371ffbbb3f30e0a16c52eeeR41).

File: Library/Homebrew/formula_installer.rb [hunk 1]
```diff
@@ -492,10 +489,14 @@ def build
...
 
     ignore_interrupts(:quietly) do # the child will receive the interrupt and marshal it back
+      socket = socket_serv.accept
+      socket.send_io write
       write.close
       data = read.read
       read.close
       Process.wait(pid)
+      socket_serv.close
+      FileUtils.rm_r File.dirname(socket_path)
```
Comment (jacknagel) [inline Library/Homebrew/formula_installer.rb]: The creation of this file should probably be at the very beginning of the method, and then this line in an ensure block at the end.

Comment (xu-cheng) [inline Library/Homebrew/formula_installer.rb]: Thanks. Updated to fix that.

File: Library/Homebrew/build.rb [hunk 2]
```diff
@@ -172,8 +172,7 @@ def fixopt f
 end
 
 begin
-  error_pipe = IO.new(ENV["HOMEBREW_ERROR_PIPE"].to_i, "w")
-  error_pipe.fcntl(Fcntl::F_SETFD, Fcntl::FD_CLOEXEC)
+  error_pipe = UNIXSocket.open(ENV["HOMEBREW_ERROR_PIPE"]).recv_io
```
Comment (jacknagel) [inline Library/Homebrew/build.rb]: This patch introduces a file descriptor leak on ruby 1.8, where close-on-exec is not the default.

Add the following lines here:

```rb
system "ls", "/dev/fd"
exit
```

On the prior implementation:

```
$ brew install tree -s
0 1 2 3 4
```

(this is the expected set of FDs for `ls` to have open.)


With this patch:

```
 $ brew install tree -s
0 1 2 3 4 5 6 7 8
```

Comment (jacknagel) [inline Library/Homebrew/build.rb]: Thinking about this code, I don't know why we are even setting this up prior to installation. It's only used in failure cases. Can we just do something like this in the rescue block (I don't know, I haven't tested it)?

```rb
UnixSocket.open(ENV["HOMEBREW_ERROR_PIPE"]) { |s| Marshal.dump(e, s.recv_io) }
```

**Comment by jacknagel** [comment]
Here's the patch I'm using which seems to solve the problem:

```diff
diff --git a/Library/Homebrew/build.rb b/Library/Homebrew/build.rb
index 98fc0ae..fd0fc05 100644
--- a/Library/Homebrew/build.rb
+++ b/Library/Homebrew/build.rb
@@ -9,6 +9,7 @@
 require "keg"
 require "extend/ENV"
 require "debrew"
+require "fcntl"
 require "socket"
 
 class Build
@@ -172,7 +173,8 @@ def fixopt f
 end
 
 begin
-  error_pipe = UNIXSocket.open(ENV["HOMEBREW_ERROR_PIPE"]).recv_io
+  error_pipe = UNIXSocket.open(ENV["HOMEBREW_ERROR_PIPE"], &:recv_io)
+  error_pipe.fcntl(Fcntl::F_SETFD, Fcntl::FD_CLOEXEC)
 
   # Invalidate the current sudo timestamp in case a build script calls sudo
   system "/usr/bin/sudo", "-k"
diff --git a/Library/Homebrew/formula_installer.rb b/Library/Homebrew/formula_installer.rb
index c394a24..a0c341a 100644
--- a/Library/Homebrew/formula_installer.rb
+++ b/Library/Homebrew/formula_installer.rb
@@ -457,7 +457,7 @@ def build_argv
 
   def build
     socket_path = "#{Dir.mktmpdir("homebrew", HOMEBREW_TEMP)}/socket"
-    socket_serv = UNIXServer.new socket_path
+    server = UNIXServer.new(socket_path)
 
     FileUtils.rm Dir["#{HOMEBREW_LOGS}/#{formula.name}/*"]
 
@@ -466,7 +466,6 @@ def build
     # 1. formulae can modify ENV, so we must ensure that each
     #    installation has a pristine ENV when it starts, forking now is
     #    the easiest way to do this
-    read, write = IO.pipe
     ENV["HOMEBREW_ERROR_PIPE"] = socket_path
 
     args = %W[
@@ -480,18 +479,18 @@ def build
 
     pid = fork do
       begin
-        read.close
+        server.close
         exec(*args)
       rescue Exception => e
         Marshal.dump(e, write)
-        write.close
         exit! 1
       end
     end
 
     ignore_interrupts(:quietly) do # the child will receive the interrupt and marshal it back
-      socket = socket_serv.accept
-      socket.send_io write
+      sock = server.accept
+      read, write = IO.pipe
+      sock.send_io(write)
       write.close
       data = read.read
       read.close
@@ -511,7 +510,7 @@ def build
     end
     raise
   ensure
-    socket_serv.close
+    server.close
     FileUtils.rm_r File.dirname(socket_path)
   end
 
```

**Comment by xu-cheng** [comment]
Thanks @jacknagel. The final fix is a little different. `write` is still required in the fork. So I added another `fcntl` call.

**Comment by jacknagel** [comment]
Ah, right.

**Comment by jacknagel** [comment]
Tested against 1.8, looks good.

**Comment by xu-cheng** [comment]
I just found this design has a fundamental flaw. `socket_serv.accept` will block the process. If some thing fails before the client socket is created, whole program will hang up. Guess I will just pass this.

**Comment by jacknagel** [comment]
Could you do something like this?

```rb
begin
  sock = server.accept_nonblock
rescue # whatever accept_nonblock raises
  if Process.waitpid(pid, Process::WNOHANG)
    # do something
  else
    retry
  end
end
```

**Comment by xu-cheng** [comment]
Let me do some test.

**Pull request closed without merging**